### DISCUSSION TITLE
How does gitbutler support multiple open branches on the same repo?

### DISCUSSION THREAD

**Discussion by c-antin** [gitbutlerapp/gitbutler#2721] [Q&A] [answered]
I just watched [Scott Chacon's FOSDEM 2024 talk on Git Tips and Tricks](https://www.youtube.com/watch?v=aolI_Rz0ZqY) and was wondering: how exactly does gitbutler support multiple open branches on the same repo?
Does it somehow open multiple workspaces with or without file checkout?
Could someone please point me to the relevant source files?

**Comment by schacon** [comment, marked as answer]
I'll try to do a longer blog post on the internals of this (been working on one for a while now), but the general idea is:

* We run a `git diff` between your working directory and your "base branch" (so, something like `origin/master`)
* Anything that is different comes back as a series of hunks in diff format (lines changed)
* We keep a list of which virtual branch each hunk "belongs" to (ownership)
* The next time we run a `git diff`, if the hunks have changed, we try to keep ownership where it was before as much as possible. New hunks we havent seen before go into a default branch (you select in the UI)
* If you drag a file or hunk to a new lane, we rewrite the ownership data so that hunk now "belongs" in the other branch
* When you commit, it writes in memory a tree that takes the "base branch" tree, applies all the owned diff hunks to it and writes out the new modified tree, which you can then push to GitHub, even though that specific tree has not technically existed on disk (similar to how `git add -i` does it.

As far as where the code is, it's all over the place, but a lot of the logic is in the `gitbutler-app/src/virtual_branches` code (like `virtual.rs`): https://github.com/gitbutlerapp/gitbutler/blob/789181fdbec89e2e040a44c61a452998fb42245c/gitbutler-app/src/virtual_branches/virtual.rs
